students/DennisLee/lesson04/trigram.py

```diff
@@ -0,0 +1,349 @@
+    return result
+
+def quote_state(previous_state, word):
+    """
+    Show whether the text currently is in the middle of a quotation that
+    needs to be ended.
+
+    :previous_state:  The state of the quotation text (**True** if it
+                      needs a closing quotation mark, **False**
+                      otherwise) before adding the next word.
+
+    :word:  The next word to be added.
+
+    :return:  The state of the quotation text (**True**/**False**) after
+              adding the next word.  It returns **None** if the state is
+              not valid (if an open quotation mark follows
+              an unended quotation mark, or if a closed quotation mark
+              is the first quotation mark or follows another closed
+              quotation mark).
+    """
+    # Code is verbose here because of confusing multi-case boolean logic
+    new_state = None
+    if word[1:-1].count('"') > 0:  
+        pass  # Disqualify interior quotation marks
+    elif previous_state == True:
+        if start_quote(word):  
+            pass  # Quotation mark can't precede next word if prev state==True
+        else:
+            new_state = not end_quote(word)
+    elif previous_state == False:
+        if not start_quote(word) and end_quote(word):
+            pass  # If prev state==False, quot marks can't only follow word
+        else:
+            new_state = start_quote(word) and not end_quote(word)
+    elif previous_state == None:
+        pass  # If prev state was already invalid, reflect that
+    return new_state
+
+def start_quote(str):
+    """
+    Show whether the current string begins with an opening quotation
+    mark.
+
+    :str:  The specified string.
+
+    :return:  **True** if the current string starts with a `"`;
+              **False** otherwise.
+    """
+    return str[0] == '"'
+
+def end_quote(str):
+    """
+    Show whether the current string ends with a closing quotation mark.
+
+    :str:  The specified string.
+
+    :return:  **True** if the current string ends with a `"`; **False**
+              otherwise.
+    """
+    return str[-1] == '"'
+
+
+if __name__ == '__main__':
+    options = {
+            '1': 'Quit', 
+            '2': 'Enter source text filename', 
+            '3': 'Enter number for random text seeding'
+    }
+    counter = 50  # Initial random number
+
+    while True:
+        # Print menu
+        print('\n\n')
+        for k, v in options.items():  
+            print(k, v)
+        
+        ui = input("\nType a menu number: ")
+        if ui == '1':  # Quit
+            print("\n\nGoodbye!")
+            break
+        elif ui == '2':  # Run the mutator with the specified filename
+            print(f'\n\nCurrent directory is {os.getcwd():s}')
+            file_name = input(
+                    '\n\nType the name of the file to mutate using trigrams: '
+                    ).strip()
+            print('\n\n', build_new_story(file_name, counter))
```

Comment:
For readability I would recommend not putting function calls into `print` statements. I wouldn't normally read `print` statements as one wouldn't expect important logic to be part of it, but some strings that get printed out.

My point is, it is easy to skip over and miss important logic piece 🙂